dotCMS/src/main/java/com/dotcms/ai/workflow/OpenAIAutoTagRunner.java

```diff
@@ -0,0 +1,247 @@
+        }
+
+
+        boolean contentNeedsSaving = false;
+        try {
+            final String response = openAIRequest(workingContentlet, contentToTag.get());
+            final List<String> tags = new ArrayList<>();
+            JSONArray tryArray = parseJsonResponse(response);
+
+            if (this.limitTags) {
+                List<String> constrainTags = findTopTags(workingContentlet);
+                tryArray.removeIf(t -> !constrainTags.contains(t.toString().toLowerCase()));
+            }
+
+            if (tryArray.isEmpty()) {
+                Logger.info(this.getClass(), "No keywords found, returning :" + workingContentlet.getTitle());
+                return;
+            }
+
+
+
+
+            tryArray.add(DOT_AI_TAGGED);
+            tryArray.forEach(t -> tags.add((String) t));
+            final Contentlet contentToSave = checkoutLatest(identifier, language, user);
+            contentNeedsSaving = setProperty(contentToSave, fieldToTry.get(), tags);
+
+
+            if (!contentNeedsSaving) {
+                Logger.warn(this.getClass(), "Nothing to save for OpenAI response: " + response);
+                return;
+            }
+            saveContentlet(contentToSave, user);
+
+
+        } catch (Exception e) {
+            final SystemMessageBuilder message = new SystemMessageBuilder().setMessage("Error:" + e.getMessage()).setLife(5000).setType(MessageType.SIMPLE_MESSAGE).setSeverity(MessageSeverity.ERROR);
+
+            SystemMessageEventUtil.getInstance().pushMessage(message.create(), List.of(user.getUserId()));
+            Logger.warn(this.getClass(), "Error:" + e.getMessage(), e);
+            throw new DotRuntimeException(e);
+        }
+    }
+
+
+
+    private String openAIRequest(Contentlet workingContentlet, String contentToTag) throws Exception {
+        Context ctx = VelocityContextFactory.getMockContext(workingContentlet, user);
+        String systemPrompt = this.limitTags ? SYSTEM_PROMPT_TAGGING_CONSTRAIN : SYSTEM_PROMPT_TAGGING_FREEFORM;
+        List<String> constrainTags = this.limitTags ? findTopTags(workingContentlet) : List.of();
+        ctx.put("constrainTags", String.join("\n", constrainTags));
+
+        final String parsedSystemPrompt = VelocityUtil.eval(systemPrompt, ctx);
+
+        final String parsedContentPrompt = VelocityUtil.eval(contentToTag, ctx);
+
+
+        JSONObject openAIResponse = CompletionsAPI.impl().prompt(parsedSystemPrompt, parsedContentPrompt, model, temperature, 2000);
+
+        return openAIResponse.getJSONArray("choices").getJSONObject(0).getJSONObject("message").getString("content");
+
+    }
+
+    private JSONArray parseJsonResponse(String responseIn) {
+        Logger.debug(this.getClass(),"---- response ----- \n" + responseIn + "\n/---- response -----");
+        String response = responseIn.replaceAll("\\R+", " ");
+        response = response.substring(response.indexOf("{"), response.lastIndexOf("}") + 1);
+        String finalResponse = response;
```

Comment:
Use `response` instead. No need for `finalResponse`.

Reply:
Done